tests/sut/kubectl.go

```diff
@@ -0,0 +1,683 @@
+you may not use this file except in compliance with the License.
+You may obtain a copy of the License at
+
+    http://www.apache.org/licenses/LICENSE-2.0
+
+Unless required by applicable law or agreed to in writing, software
+distributed under the License is distributed on an "AS IS" BASIS,
+WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+See the License for the specific language governing permissions and
+limitations under the License.
+*/
+
+package sut
+
+import (
+	"bytes"
+	"encoding/json"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"runtime/debug"
+	"strings"
+	"time"
+
+	"github.com/pkg/errors"
+	"go.uber.org/zap"
+	"gopkg.in/yaml.v2"
+	v1 "k8s.io/api/core/v1"
+	"k8s.io/apimachinery/pkg/util/wait"
+)
+
+const (
+	helmCmd    = "helm"
+	kubeCtlCmd = "kubectl"
+)
+
+// Kubectl is used as a command to test e2e tests
+type Kubectl struct {
+	Log          *zap.SugaredLogger
+	Namespace    string
+	PollTimeout  time.Duration
+	pollInterval time.Duration
+}
+
+// NewKubectl returns a new CfOperatorkubectl command
+func NewKubectl() *Kubectl {
+	return &Kubectl{
+		Namespace:    "",
+		PollTimeout:  300 * time.Second,
+		pollInterval: 500 * time.Millisecond,
+	}
+}
+
+// RunCommandWithCheckString runs the command specified helper in the container
+func (k *Kubectl) RunCommandWithCheckString(namespace string, podName string, commandInPod string, result string) error {
+	out, err := runBinary(kubeCtlCmd, "--namespace", namespace, "exec", podName, "--", "sh", "-c", commandInPod)
+	if err != nil {
+		return err
+	}
+	if strings.Contains(string(out), result) {
+		return nil
+	}
+	return errors.Errorf("'%s' not found in output '%s'", result, string(out))
+}
+
+// GetPodNames returns the names of the pods matching the selector
+func (k *Kubectl) GetPodNames(namespace string, selector string) ([]string, error) {
+	out, err := runBinary(kubeCtlCmd, "--namespace", namespace, "get", "pod",
+		"-l", selector,
+		"-o", "jsonpath={.items[*].metadata.name}")
+	if err != nil {
+		return []string{}, err
+	}
+	names := strings.Split(string(out), " ")
+	return names, nil
+}
+
+// WaitForPod blocks until the pod is available. It fails after the timeout.
+func (k *Kubectl) WaitForPod(namespace string, labelName string, podName string) error {
+	return wait.PollImmediate(k.pollInterval, k.PollTimeout, func() (bool, error) {
+		return k.PodExists(namespace, labelName, podName)
+	})
+}
+
+// WaitForPodDelete blocks until the pod is available. It fails after the timeout.
+func (k *Kubectl) WaitForPodDelete(namespace string, podName string) error {
+	return wait.PollImmediate(k.pollInterval, k.PollTimeout, func() (bool, error) {
+		return k.checkPodDeleted(namespace, podName)
+	})
+}
+
+// checkPodTerminateLabelFilter checks is the pod status is terminated
+func (k *Kubectl) checkPodDeleted(namespace string, name string) (bool, error) {
+	out, err := runBinary(kubeCtlCmd, "--namespace", namespace, "get", "pod", name)
```

Comment:
isnt there a golang way of getting this info with a client instead of forking to binaries?

Reply:
yes but wouldn't be anymore an e2e test..plus we would end up pinning deps that we don't really need. Here we can later have a matrix for kubernetes versions indipendent from the code